zebra/zebra_nhg.c

```diff
@@ -99,31 +99,53 @@ nhg_connected_tree_root(struct nhg_connected_tree_head *head)
 	return nhg_connected_tree_first(head);
 }
 
-void nhg_connected_tree_del_nhe(struct nhg_connected_tree_head *head,
-				struct nhg_hash_entry *depend)
+struct nhg_hash_entry *
+nhg_connected_tree_del_nhe(struct nhg_connected_tree_head *head,
+			   struct nhg_hash_entry *depend)
 {
 	struct nhg_connected lookup = {};
 	struct nhg_connected *remove = NULL;
+	struct nhg_hash_entry *removed_nhe;
 
 	lookup.nhe = depend;
 
 	/* Lookup to find the element, then remove it */
 	remove = nhg_connected_tree_find(head, &lookup);
 	remove = nhg_connected_tree_del(head, remove);
 
-	if (remove)
+	if (remove) {
+
+		removed_nhe = remove->nhe;
 		nhg_connected_free(remove);
+		return removed_nhe;
+	}
+
+	return NULL;
 }
 
-void nhg_connected_tree_add_nhe(struct nhg_connected_tree_head *head,
-				struct nhg_hash_entry *depend)
+/* Assuming UNIQUE RB tree. If this changes, assumptions here about
+ * insertion need to change.
+ */
+struct nhg_hash_entry *
+nhg_connected_tree_add_nhe(struct nhg_connected_tree_head *head,
+			   struct nhg_hash_entry *depend)
 {
 	struct nhg_connected *new = NULL;
 
 	new = nhg_connected_new(depend);
 
-	if (new)
-		nhg_connected_tree_add(head, new);
+	/* On success, NULL should be returned from the
+	 * RB code.
+	 */
+	if (new && (nhg_connected_tree_add(head, new) == NULL))
+		return NULL;
+
+	/* If it wasn't successful, it must be a duplciate. The kernel
```

Comment:
typo: "duplicate"

I'd rather not make the claim about "the kernel" - it'd be clearer to say "we enforce this property about the 'connected' tree: ..."

Reply:
will change